controller/resource.go

```diff
@@ -0,0 +1,240 @@
+	"github.com/aws/aws-sdk-go/service/ecs/ecsiface"
+	"github.com/aws/aws-sdk-go/service/secretsmanager"
+	"github.com/aws/aws-sdk-go/service/secretsmanager/secretsmanageriface"
+	"github.com/hashicorp/consul/api"
+	"github.com/hashicorp/go-hclog"
+)
+
+const meshTag = "consul.hashicorp.com/mesh"
+
+// ResourceLister is an interface for listing Resources.
+type ResourceLister interface {
+	List() ([]Resource, error)
+}
+
+// Resource is a generic type that needs to be reconciled by the Controller.
+// It offers Upsert and Delete functions to reconcile itself with an external state.
+type Resource interface {
+	ID() (string, error)
+	Upsert() error
+	Delete() error
+}
+
+// TaskLister is an implementation of ResourceLister that lists ECS tasks.
+type TaskLister struct {
+	// ECSClient is the AWS ECS client to be used by the TaskLister.
+	ECSClient ecsiface.ECSAPI
+	// SecretsManagerClient is the AWS Secrets Manager client to be used by the TaskLister.
+	SecretsManagerClient secretsmanageriface.SecretsManagerAPI
+	// ConsulClient is the Consul client to be used by the TaskLister.
+	// TaskLister doesn't need to talk to Consul, but it passes this client
+	// to each Resource it creates.
+	ConsulClient *api.Client
+
+	// Cluster is the name or the ARN of the ECS cluster.
+	Cluster string
+	// SecretPrefix is the prefix to determine names of resources in Consul or AWS.
+	SecretPrefix string
+
+	// Log is the logger for the TaskLister.
+	Log hclog.Logger
+}
+
+// List lists all tasks for the Cluster.
+func (t TaskLister) List() ([]Resource, error) {
+	var resources []Resource
+	// nextToken is to handle paginated responses from AWS.
+	var nextToken *string
+
+	// This isn't an infinite loop, instead this is a "do while" loop
+	// because we'll break out of it as soon as nextToken is nil.
+	for {
+		taskListOutput, err := t.ECSClient.ListTasks(&ecs.ListTasksInput{
+			Cluster:   aws.String(t.Cluster),
+			NextToken: nextToken,
+		})
+		if err != nil {
+			return nil, fmt.Errorf("listing tasks: %w", err)
+		}
+		nextToken = taskListOutput.NextToken
+
+		tasks, err := t.ECSClient.DescribeTasks(&ecs.DescribeTasksInput{
+			Cluster: aws.String(t.Cluster),
+			Tasks:   taskListOutput.TaskArns,
+			Include: []*string{aws.String("TAGS")},
+		})
+		if err != nil {
+			return nil, fmt.Errorf("describing tasks: %w", err)
+		}
+		for _, task := range tasks.Tasks {
+			resources = append(resources, &Task{
+				SecretsManagerClient: t.SecretsManagerClient,
+				ConsulClient:         t.ConsulClient,
+				Cluster:              t.Cluster,
+				Log:                  t.Log,
+				SecretPrefix:         t.SecretPrefix,
+				Task:                 task,
+			})
+		}
+		if nextToken == nil {
+			break
+		}
+	}
+	return resources, nil
+}
+
+type Task struct {
+	SecretsManagerClient secretsmanageriface.SecretsManagerAPI
+	ConsulClient         *api.Client
+
+	Cluster      string
+	SecretPrefix string
+	Task         *ecs.Task
+
+	Log hclog.Logger
+}
```

Comment:
This set of fields kinda confuses me to be on `Task`. I see we need them because we're caling `task.Upsert()` but I'm curious why we didn't go with an interface of `TaskController.Upsert(*ecs.Task)` and that `TaskController` struct would have all the `SecretsManagerClient`, `ConsulClient`, etc fields.

For example our config entry ctrls in k8s have this interface where they take in a `req` which is essentially the `task`:

```go
func (r *ConfigEntryController) ReconcileEntry(ctx context.Context, crdCtrl Controller, req ctrl.Request, configEntry common.ConfigEntryResource)
```

Reply:
So then I guess `controller` would be more the ControllerManager in my head and then we'd write an `ACLController` that has interfaces:

```go
Upsert(*ecs.Task)
Delete(*ecs.Task)
```

Reply:
Yeah, I see what you mean about the fields. The idea here was to have a generic controller implementation that can reconcile any kind of resource. Then it'd be up to the resource's implementation to decide what it means for it to upsert or to delete itself. For that, it needs to have external store parameters, like the clients etc so that it can do those things. This approach was probably more inspired by the [`ServiceResource`](https://github.com/hashicorp/consul-k8s/blob/c79730ba4f0e25b8330950d7ab575e4e3cfc1ef0/control-plane/catalog/to-consul/resource.go#L51) we have in the k8s sync-catalog rather than the `controller-runtime` type controllers. `ServiceResource` similarly is a type that has fields it needs to "reconcile" itself. The reason I went with this approach is that it seemed more intuitive to me for this use case. Meaning it seems more intuitive to say `resource.Upsert` rather than `controller.Upsert`, but this might be just a personal preference.

I think what you're suggesting could also work. Other than what I said above, I don't have a super-strong opinion on which way is better. If we do want to change it, I'd do it in a separate PR.

Reply:
Agreed on separate PR. I prefer the Kubernetes controller approach but I think that's also just a personal preference thing. @pglass something to think about after this release.